File: java/src/main/java/S059.java
import com.google.common.base.CharMatcher;
import com.google.common.base.Splitter;

import java.util.stream.IntStream;

public class S059 {

    private static final int KEY_SIZE = 3;
    private static final CharMatcher LETTER_MATCHER =
            CharMatcher.inRange('a', 'z').or(CharMatcher.inRange('A', 'Z'));

    public static long solve() {
        int[] cipher = parseInput(PeResources.getResource("059.txt"));
        int[] keys = new int[]{bestKey(cipher, 0), bestKey(cipher, 1), bestKey(cipher, 2)};
        return IntStream.range(0, cipher.length).map(i -> keys[i % 3] ^ cipher[i]).sum();
    }

    private static int bestKey(int[] cipher, int skip) {
        int bestKey = -1;
        double bestScore = -1;
        for (int i = 'a'; i <= 'z'; i++) {
            double score = tryKey(cipher, i, skip);
            if (score > bestScore) {
                bestKey = i;
                bestScore = score;
            }
        }
        return bestKey;
    }

    private static double tryKey(int[] cipher, int key, int skip) {
        int hits = 0;
        for (int i = skip; i < cipher.length; i += KEY_SIZE) {
            if (LETTER_MATCHER.matches((char) (cipher[i] ^ key))) {
                hits++;
            }
        }
        return ((double) hits) / cipher.length;
    }

    private static int[] parseInput(String input) {
        String cleaned = CharMatcher.whitespace().removeFrom(input);
        return Splitter.on(',').splitToList(cleaned).stream()
                .mapToInt(Integer::parseInt)
                .toArray();
    }

}

File: cpp/S059.cpp
#include <iostream>
#include "files.cpp"
#include "strings.cpp"

using namespace std;

static const int KEY_SIZE = 3;

inline bool is_letter(char c) {
    return (c >= 'a' && c <= 'z')
        || (c >= 'A' && c <= 'Z');
}

inline int ascii_sum_of_decrypted(const vector<char>& values, string key) {
    int sum = 0;
    for (int i = 0; i < values.size(); i++) {
        sum += key[i % KEY_SIZE] ^ values[i];
    }
    return sum;
}

inline int score_for_key(const vector<char>& values, char key, int skip) {
    int idx = skip;
    int score = 0;
    while (idx < values.size()) {
        score += is_letter(values[idx] ^ key);
        idx += KEY_SIZE;
    }
    return score;
}

int main() {
    vector<string> letters;
    split(read_file_as_string("inputs/059.txt"), ',', letters);
    vector<char> values;
    for (string letter : letters) {
        values.push_back(stoi(letter));
    }
    string key;
    for (int i = 0; i < KEY_SIZE; i++) {
        char best_key = -1;
        int best_score = -1;
        for (char c = 'a'; c <= 'z'; c++) {
            int score = score_for_key(values, c, i);
            if (score > best_score) {
                best_score = score;
                best_key = c;
            }
        }
        key += best_key;
    }
    cout << ascii_sum_of_decrypted(values, key) << endl;
    return 0;
}
